evaltools/data/estimatecvap.py

```diff
@@ -0,0 +1,258 @@
+
+    Returns: 
+       `base` geometries with 2019 CVAP-weighted 2020 CVAP estimates attached.
+    """
+    if geometry10 not in {"block group", "tract"}:
+        print(f"Requested geometry \"{geometry10}\" is not allowed; loading tracts.")
+        geometry10 = "tract"
+
+    # Grab ACS and CVAP special-tab data, and make sure our triples are correct
+    cvap_geoid = "TRACT10" if geometry10 == "tract" else "BLOCKGROUP10"
+    acs_source = acs5(state, geometry10)
+    cvap_source = cvap(state, geometry10)
+
+    # Validate the columns passed, issuing user warnings when it's inadvisable
+    # to estimate CVAP given the passed columns.
+    for (cvap10, vap10, vap20) in groups:
+        # If any of the CVAP columns passed correspond to columns which tabulate
+        # people of multiple races, notify the user that there isn't an appropriate
+        # 2019 VAP column to match them against.
+        if any(substring in cvap10 for substring in {"AIW", "AW", "BW", "AIB"}):
+            print(
+                f"Warning: Estimating CVAP among {cvap10} is not advisable, since "
+                "there isn't a reasonable VAP column from which to construct _CVAP "
+                "/ _VAP rates (because you seem to be combining two racial groups)."
+            )
+
+        # If the CVAP or ACS5 columns passed aren't present in the set of possible
+        # columns, raise an error.
+        if not (cvap10 in acs_source or cvap10 in cvap_source):
+            possible_columns = set(acs_source).union(set(cvap_source))
+            raise ValueError(
+                f"Your CVAP column '{cvap10}' must be contained in either the ACS "
+                f"or Special Tab columns: {possible_columns}"
+            )
+        
+        if not vap10 in acs_source:
+            raise ValueError(
+                f"Your old VAP column '{vap10}' must be contained in the ACS "
+                "columns: {set(acs_source)}"
+            )
+
+        # If the VAP20 column passed doesn't exist on the user-provided geometries,
+        # raise an error.
+        if not vap20 in base:
+            raise ValueError(
+                f"Your new VAP column '{vap20}' must be contained in your base " +
+                f"dataframe: {set(base)}"
+            )
+    
+    # Remove ACS 5 columns that overlap with special-tab ones.
+    non_overlaps = list(set(acs_source).difference(set(cvap_source)))
+    acs_source = acs_source[[cvap_geoid] + non_overlaps]
+    source = cvap_source.merge(acs_source, on=cvap_geoid)
+
+    # Get the right columns from the base geometry, and map the base geometries
+    # to the units with CVAP data. Apparently dropping bad columns by using slicing
+    # incudes a SettingWithCopy warning, so we're just dropping using .drop()
+    # instead.
+    correct = ["geometry"] + [col for col in list(base) if any(sub in col for sub in ["POP", "VAP", "GEOID"])]
+    bads = list(set(base) - set(correct))
+    pared = base.drop(bads, axis=1)
```

Comment:
Slicing induces a pandas `SettingWithCopyWarning`, so we `.drop()` bad columns instead.